File: Python/Task27/n27891.py
# https://inf-ege.sdamgia.ru/problem?id=27891

import time

def simpleAlgo(filename):
	file = open(filename, "r")
	n = int(file.readline())
	a = []
	for line in file:
		a.append(int(line))
	x = 0
	for i, a_i in enumerate(a):
		for a_j in a[i + 1:]:
			t = a_i * a_j;
			if t % 14 == 0:
				x = max(t, x)
	return x

def onePassAlgo(filename):
	file = open(filename, "r")
	n = int(file.readline())
	max2 = max7 = max14 = maxOther = 0
	for i in range(0, n):
		t = int(file.readline())
		if t % 7 == 0 and t % 2 != 0:
			max7 = max(t, max7)
		if t % 2 == 0 and t % 7 != 0:
			max2 = max(t, max2)
		if t % 14 == 0 and t > max14:
			maxOther = max(max14, maxOther)
			max14 = t
		else:
			maxOther = max(t, maxOther)
	if max2 * max7 > max14 * maxOther:
		return max2 * max7
	else:
		return max14 * maxOther

filename = "n27891_B.txt"

print("Решение в один проход через вычисление количества чисел, делящихся на 2, 7 и 14")
start_time = time.time()
x = onePassAlgo(filename)
print("Execution time: %s seconds" % (time.time() - start_time))
print("X =", x, "\n")

print("Решение простым перебором")
start_time = time.time()
x = simpleAlgo(filename)
print("Execution time: %s seconds" % (time.time() - start_time))
print("X =", x)

File: C++/Task27/n27891/n27891.cpp
// https://inf-ege.sdamgia.ru/problem?id=27891

#include <iostream>
#include <fstream>
#include <chrono>

int simpleAlgo(const char* filename)
{
	std::fstream file(filename, std::ios_base::in);
	int n;
	file >> n;
	int* a = new int[n];
	for (int i = 0; i < n; i++)
	{
		file >> a[i];
	}
	int x = 0;
	for (int i = 0; i < n - 1; i++)
	{
		for (int j = i + 1; j < n; j++)
		{
			int t = a[i] * a[j];
			if (t % 14 == 0)
			{
				x = std::max(t, x);
			}
		}
	}
	return x;
}

int onePassAlgo(const char* filename)
{
	std::fstream file(filename, std::ios_base::in);
	int n;
	file >> n;
	int max2 = 0, max7 = 0, max14 = 0, maxOther = 0, t;
	for (int i = 0; i < n; i++)
	{
		file >> t;
		if (t % 7 == 0 && t % 2 != 0)
		{
			max7 = std::max(t, max7);
		}
		if (t % 2 == 0 && t % 7 != 0)
		{
			max2 = std::max(t, max2);
		}
		if (t % 14 == 0 && t > max14)
		{
			maxOther = std::max(max14, maxOther);
			max14 = t;
		}
		else
		{
			maxOther = std::max(t, maxOther);
		}
	}
	if (max2 * max7 > max14 * maxOther)
	{
		return max2 * max7;
	}
	else
	{
		return max14 * maxOther;
	}
}

using std::chrono::high_resolution_clock;
using std::chrono::duration_cast;
using std::chrono::duration;
using std::chrono::milliseconds;
using std::chrono::steady_clock;
using std::cout;
using std::endl;

int main()
{
	const char* filename = ".\\Data\\27-B_2.txt";
	{
		cout << "Решение в один проход через вычисление количества чисел, делящихся на 2, 7 и 14" << endl;
		steady_clock::time_point time_start = high_resolution_clock::now();
		int x = onePassAlgo(filename);
		steady_clock::time_point time_end = high_resolution_clock::now();
		milliseconds ms = duration_cast<milliseconds>(time_end - time_start);
		cout << "Execution time: " << ms.count() / 1000.0 << " seconds" << endl;
		cout << "X = " << x << endl << endl;
	}

	{
		cout << "Решение простым перебором" << endl;
		steady_clock::time_point time_start = high_resolution_clock::now();
		int x = simpleAlgo(filename);
		steady_clock::time_point time_end = high_resolution_clock::now();
		milliseconds ms = duration_cast<milliseconds>(time_end - time_start);
		cout << "Execution time: " << ms.count() / 1000.0 << " seconds" << endl;
		cout << "X = " << x << endl;
	}
	
	return 0;
}